Model/Queue.php

```diff
@@ -66,56 +87,11 @@ public function runCron()
         $this->run($nbJobs);
     }
 
-    public function run($limit)
+    public function run($maxJobs)
     {
-        $full_reindex = ($limit === -1);
-        $limit = $full_reindex ? 1 : $limit;
-
-        $element_count = 0;
-        $jobs = [];
-        $offset = 0;
-        $max_size = $this->configHelper->getNumberOfElementByPage() * $limit;
-
-        while ($element_count < $max_size) {
-            $data = $this->db->query($this->db->select()->from($this->table, '*')->where('pid IS NULL')
-                                              ->order(['job_id'])->limit($limit, $limit * $offset));
-            $data = $data->fetchAll();
-
-            $offset++;
-
-            if (count($data) <= 0) {
-                break;
-            }
-
-            foreach ($data as $job) {
-                $job_size = (int) $job['data_size'];
-
-                if ($element_count + $job_size <= $max_size || empty($jobs)) {
-                    $jobs[] = $job;
-                    $element_count += $job_size;
-                } else {
-                    break 2;
-                }
-            }
-        }
-
-        if (count($jobs) <= 0) {
-            return;
-        }
-
-        $first_id = $jobs[0]['job_id'];
-        $last_id = $jobs[count($jobs) - 1]['job_id'];
-
         $pid = getmypid();
 
-        // Reserve all new jobs since last run
-        $this->db->query("UPDATE {$this->db->quoteIdentifier($this->table, true)} SET pid = " . $pid . ' WHERE job_id >= ' . $first_id . " AND job_id <= $last_id");
-
-        foreach ($jobs as &$job) {
-            $job['data'] = json_decode($job['data'], true);
-        }
-
-        $jobs = $this->sortAndMergeJob($jobs);
+        $jobs = $this->getJobs($maxJobs, $pid);
```

Comment:
Maybe we should get the job outside this loop (loop because run is called recursively). Jobs added in the queue while running a full reindex won't interfere. What do you think?